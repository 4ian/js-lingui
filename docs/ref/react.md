# Reference: React Components

- Providers
  - [I18nProvider](#i18nprovider)
  - [WithI18n](#withi18n)

|| Example MessageFormat
---        | --
[Trans](#trans) | `My name is {name}` 
**Categories** ||
[Plural](#plural) | `{count, plural, one {# book} other {# books}}`
[SelectOrdinal](#selectordinal) | `{count, selectordinal, one {#st} two {#nd} few {#rd} other {#th}}`
[Select](#select) | `{gender, select, male {His} female {Her} other {Their}}`
**Formats** ||
[DateFormat](#dateformat) | `{value, date, short}`
[NumberFormat](#numberformat) | `{value, number, percent}`
  
## General concepts
  
### Rendering in components

All i18n components render translation inside `<span>` tag. This tag can be customized using two props:

Prop name | Type | Description 
--- | --- | --- |
`className` | string | Class name to be added to `<span>` element
`render` | React.Element, React.Class string | Custom wrapper element to render translation

`className` is ignored, `render` is set.

When `render` is **React.Element** or **string** (built-in tags), it is cloned with `translation` as children:

```jsx
// built-in tags
<Trans render="h1">Heading</Trans>

// custom elements
<Trans render={<Link to="/docs" />}>Link to docs</Trans>
```

Using React.Component (or stateless component) in `render` prop is useful to get more control over the rendering of translation. Component passed to `render` receive translation/value in `translation` prop.

```jsx
// custom component
<Trans render={
  ({ translation }) => <Icon label={translation} />
}>
    Sign in
</Trans>

// renders as
// <Icon label="Sign in" />
```

### Variables inside components

**:warning: Only simple variables are supported inside [Trans](#trans) component and as a `value` prop in other i18n components.**

```js
// This is valid example
const name = "Arthur"

// This is valid
<Trans>Hello {name}</Trans>
```

It's not possible to use constants, function calls or object properties. 

Any of following examples won't work:

```js
const messages = [
  "Hello",
  "World"
]

// These are invalid
<Trans>{messages.length}</Trans>
<Trans>{isEmpty(messages)}</Trans>
<Trans>The answer is {42}</Trans>
<DateFormat value={new Date()}
```

## Components

### Trans

This is the main component for translation. It supports variables and components inside messages.

Message ID is either automatically generated in babel plugin or it's possible to manually override it in `id` prop.

Prop name | Type | Description
--- | --- | ---
`id` | string | Override auto-generated message ID

```jsx
// basic use case
<Trans>Hello World</Trans>
```

```jsx
// custom message ID
<Trans id="msg.hello">Hello World</Trans>
```

```jsx
// variable interpolation
const name = "Fred"
<Trans>My name is {name}</Trans>
```

```jsx
// inline components
<Trans>See the <Link to="/more">description</Link> below.</Trans>
```

### Trans (without babel plugin)

It's also possible to use `Trans` component without babel plugin. In fact, it's the only i18n component you'll need if you decide to go without babel plugin.

**Note:** If you're using `babel-preset-lingui`, you don't need to use any of these props. It will be set automatically. 

Prop name | Type | Description
--- | --- | ---
`id` | string | Message ID
`defaults` | string | Default translation
`values` | Object | Variables for interpolation
`components` | Array | Components used in message
`formats` | Object | Custom formats used in message

#### Examples

```jsx
<Trans id="Hello World" />
```

```jsx
<Trans 
  id="Hello {name}" 
  values={{ name: 'Arthur' }}
/>
```

```jsx
// number of tag corresponds to index in `components` prop
<Trans 
  id="Read <0>Description</0> below." 
  components={[<Link to="/docs" />]} 
/>
```

```jsx
<Trans 
  id="Today is {today, date, short_date}"
  values={{ today: new Date() }}
  formats={{ 
    short_date: {
      year: "numberic",
      month: "long",
      day: "numeric"
    }
  }} 
/>
```


### Plural

MessageFormat: `{arg, plural, ...forms}`

Prop name | Type | Description
--- | --- | ---
`value` | number | Override auto-generated message ID
`offset` | number | Offsets plural forms but doesn't affect exact matches
`zero`  | string | Form for empty `value`
`one`   | string | *Singular* form
`two`   | string | *Dual* form
`few`   | string | *Paucal* form
`many`  | string | *Plural* form
`other` | string | (required) general *plural* form
`_0`, `_1`, ... | string | Exact match form, correspond to `=N` rule

See [Language Plural Rules](http://www.unicode.org/cldr/charts/latest/supplemental/language_plural_rules.html) overview.

Based on language, the `value` is mapped to specific plural form. If such plural form isn't defined, the `other` form is used. `#` character inside message is used as a placeholder for `value`:

```jsx
const count = 42
// renders as '42 books'
<Plural 
    value={count}
    one="# book"
    other="# books"
/>
```

It's also possible to use exact matches. This is common used in combination with `offset` prop. `offset` doesn't affect `value` for exact matches, only plural forms:

```jsx
const count = 42
<Plural 
    value={count}
    offset={1}
    // when value = 0
    _0="Nobody arrived"
    
    // when value = 1
    _1="Only you arrived"
    
    // when value = 2
    // value - offset = 1 -> `one` plural form
    one="You and # other guest"
    
    // when value >= 3
    other="You and # other guests"
/>
```

### Select

MessageFormat: `{arg, select, ...forms}`

Prop name | Type | Description
--- | --- | ---
`value` | number | Override auto-generated message ID
`other` | number | (required) Default, catch-all form

This component selects appropriate form based on content of `value`. It behaves like an `switch` statement. `other` prop is used when no prop matches `value`:

```jsx
// gender = "female"      -> `Her book`
// gender = "male"        -> `His book`
// gender = "unspecified" -> `Their book`
<Select 
    value={gender}
    male="His book"
    female="Her book"
    other="Their books"
/>
```

### SelectOrdinal

MessageFormat: `{arg, selectordinal, ...forms}`

Prop name | Type | Description
--- | --- | ---
`value` | number | Override auto-generated message ID
`offset` | number | Offsets plural forms but doesn't affect exact matches
`zero`  | string | Form for empty `value`
`one`   | string | *Singular* form
`two`   | string | *Dual* form
`few`   | string | *Paucal* form
`many`  | string | *Plural* form
`other` | string | (required) general *plural* form
`_0`, `_1`, ... | string | Exact match form, correspond to `=N` rule

See [Language Plural Rules](http://www.unicode.org/cldr/charts/latest/supplemental/language_plural_rules.html) overview.

This component is equivalent to [Plural](#plural). The only difference is that it uses **ordinal** plural forms, instead of **cardinal** ones.

```jsx
<SelectOrdinal
    value={count}
    one="#st"
    two="#nd"
    few="#rd"
    other="#th"
/>
```

### DateFormat

MessageFormat: `{arg, date, format}`

Prop name | Type | Description
--- | --- | ---
`value` | number | Date to be formatted
`format`  | string or Object | Date format passed as options to [Intl.DateTimeFormat](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/DateTimeFormat)

This component is a wrapper around [Intl.DateTimeFormat](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/DateTimeFormat).

```jsx
const now = new Date()
// default language format
<DateFormat value={now} />
```

```jsx
const now = new Date()
// custom format
<DateFormat value={now} format={{
    year: "numberic",
    month: "long",
    day: "numeric"
}}/>
```

### NumberFormat

MessageFormat: `{arg, number, format}`

Prop name | Type | Description
--- | --- | ---
`value` | number | Number to be formatted
`format`  | string or Object | Number format passed as options to [Intl.NumberFormat](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/NumberFormat)

This component is a wrapper around [Intl.NumberFormat](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/NumberFormat).

```jsx
const num = 0.42
// default language format
<NumberFormat value={num} />
```

```jsx
const amount = 3.14
// custom format
<NumberFormat value={amount} format={{
    style: "currency",
    currency: 'EUR',
    minimumFractionDigits: 2
}}/>
```

## Providers

Message catalogs and active language are passed to the context in 
[I18nProvider](#i18nprovider). However, context should never be accessed 
directly. [WithI18n](#withi18n) hoc passes i18n prop to wrapped component
and shadows all implementation details:

### I18nProvider

Prop name | Type | Description
--- | --- | ---
`language` | string | Active language
`messages` | object | Message catalog

`messages` must be in format: `{ [language]: messageCatalog }`. This component should live above all i18n components. The good place is top-level application component. However, if the `language` is stored in the `redux` store, this component should be inserted below `react-redux/Provider`.

```jsx
import React from 'react'
import { Provider, connect } from 'react-redux'
import { I18nProvider } from 'lingui-react'

const App = connect(state => ({ language: state.language }))(
    ({ language} ) => {
        const messages = require(`locales/${language}.json`)
        return (
            <I18nProvider language={language} messages={{ [language]: messages }}>
               // the rest of app
            </I18nProvider>,
        )
    }
)
```

### WithI18n

`WithI18n([ options ])`

Options | Description
--- | ---
`withRef` | Returns reference to wrapped instance in `getWrappedInstance`

This HOC injects `i18n` prop to wrapped component. It's useful when wrapped component needs to access `lingui-i18n` API for plain text translations.

```jsx
import React from 'react'
import { Trans, WithI18n } from 'lingui-react'

// WithI18n injects `i18n` prop 
const LogoutIcon = WithI18n()(
    ({ i18n }) => <Icon name="turn-off" aria-label={i18n.t`Log out`}/>
)
```
