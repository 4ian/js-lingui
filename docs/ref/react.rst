*****************************************
@lingui/react - React components for i18n
*****************************************

Components from ``@lingui/react`` wrap the vanilla JS API from
:doc:`@lingui/core <core>`. React components handle changes of active language and
interpolated variables better than low-level API and also take care of re-rendering when
wrapped inside pure components.

Installation
============

.. code-block:: sh

   yarn add @lingui/react
   # npm install --save @lingui/react

Reference
=========

Trans
-----

.. component:: Trans

   :prop string id: ID of message to load from catalog and render
   :prop string|React.ElementType render:
   :prop Object values:
   :prop Array<Element> components:
   :prop Object formats:

This is the main and most-used component for translation. It supports
variables and components inside messages. Usage of this component depends on
whether or not you're using jsLingui Babel plugins.

Each message is identified by **message ID**.
``babel-plugin-lingui-transform-react`` automatically generates message ID from
contents of :component:`Trans` component, but it's possible to provide custom
message ID by setting the `id` prop.

.. code-block:: jsx

   <Trans>Hello World</Trans>;

   // custom message ID
   <Trans id="msg.hello">Hello World</Trans>;

   // variable interpolation
   const name = "Fred";
   <Trans>My name is {name}</Trans>;

    // inline components
    <Trans>See the <Link to="/more">description</Link> below.</Trans>;

It's also possible to use :component:`Trans` component without babel plugin. In
fact, it's the only i18n component you'll need if you decide to go without the babel plugin.

.. code-block:: jsx

   <Trans id="Hello World" />;

   <Trans
     id="Hello {name}"
     values={{ name: 'Arthur' }}
   />;

   // number of tag corresponds to index in `components` prop
   <Trans
     id="Read <0>Description</0> below."
     components={[<Link to="/docs" />]}
   />;

   <Trans
     id="Today is {today, date, short_date}"
     values={{ today: new Date() }}
     formats={{
       short_date: {
         year: "numeric",
         month: "long",
         day: "numeric"
       }
     }}
   />;

DateFormat
----------

.. component:: DateFormat

   :prop string|Date value: Date to be formatted
   :prop string|Object format: Date format passed as options to `Intl.DateTimeFormat`_.

:component:`DateFormat` component is a wrapper around `Intl.DateTimeFormat`_.
It takes a date object or a date string as a ``value`` prop and formats it using
``format`` options. ``format`` prop supports the same options as `Intl.DateTimeFormat`_:

.. code-block:: jsx

   // date as a string
   <DateFormat value="2018-07-23" />;

   const now = new Date();
   // default language format
   <DateFormat value={now} />;

   const now = new Date();
   // custom format
   <DateFormat value={now} format={{
       year: "numeric",
       month: "long",
       day: "numeric"
   }} />;

NumberFormat
------------

.. component:: NumberFormat

   :prop number value: Number to be formatted
   :prop string|Object format: Number format passed as options to `Intl.NumberFormat`_

:component:`NumberFormat` component is a wrapper around
`Intl.NumberFormat_. It takes a number as a ``value`` prop
and formats it using ``format`` options. ``format`` prop supports the same
options as `Intl.NumberFormat`_:

.. code-block:: jsx

   const num = 0.42;
   // default language format
   <NumberFormat value={num} />;

   const amount = 3.14;
   // custom format
   <NumberFormat value={amount} format={{
       style: 'currency',
       currency: 'EUR',
       minimumFractionDigits: 2
   }} />;

Providers
=========

Message catalogs and the active language are passed to the context in
:component:`I18nProvider`. However, context should never be accessed
directly. The :js:func:`withI18n` high-order component passes ``i18n`` prop
down to wrapped component and shadows all implementation details.

I18nProvider
------------

.. component:: I18nProvider

   :prop string language: Active language
   :prop string|string[] locales: List of locales used for date/number formatting. Defaults to active language.
   :prop object catalogs: Message catalogs
   :prop React.Element|React.Class|string defaultRender: Default element to render translation
   :prop string|Function missing: Custom message to be returned when translation is missing

``defaultRender`` has the same meaning as ``render`` in other i18n
components. :ref:`Rendering of translations <rendering-translations>` is explained
at the beginning of this document.

``language`` sets the active language and loads corresponding message catalog.
``locales`` are used for date/number formatting for countries or regions which use
different formats for the same language (e.g. arabic numerals have several
representations).

``missing`` is used as a default translation when translation is missing. It might
be also a function, which is called with language and message ID. This is useful
for debugging:

.. code-block:: jsx

   import React from 'react';
   import { I18nProvider } from '@lingui/react';

   const App = ({ language} ) => {
        return (
            <I18nProvider language={language} missing="ðŸš¨">
               {/* This will render as ðŸš¨*/}
               <Trans id="missing translation" />
            </I18nProvider>
        );
   }

``catalogs`` is a type of ``Catalogs``:

.. code-block:: jsx

   // One catalog per language
   type Catalogs = {
     [language: string]: Catalog
   }

   // Catalog contains messages and language data (i.e: plurals)
   type Catalog = {
     messages: Messages,
     languageData?: {
       plurals: Function
     }
   }

   // Message is either function (compiled message) or string
   type Messages = {
     [messageId: string]: string | Function
   }

This component should live above all i18n components. A good place is as a
top-level application component. However, if the ``language`` is stored in a
``redux`` store, this component should be inserted below ``react-redux/Provider``:

.. code-block:: jsx

   import React from 'react';
   import { I18nProvider } from '@lingui/react';

   const App = ({ language} ) => {
        const catalog = require(`locales/${language}.js`);

        return (
            <I18nProvider language={language} catalogs={{ [language]: catalog }}>
               // rest of the app
            </I18nProvider>
        );
   }

I18n
----

.. component:: I18n

   :prop bool update: Subscribe to catalog and activate language updates

:component:`I18n` injects ``i18n`` object and ``i18nHash`` to child component, which
may be lambda component, regular component or React element. This pattern is
known as `render prop component <https://reactjs.org/docs/render-props.html>`_.

If want to use ``i18n`` object in instance or lifecycle methods, consider using
:js:func:`withI18n` high-order component.

``i18nHash`` is useful when rendering pure components or elements as it contains
hash of active language and catalogs. Instead of comparing ``i18n`` object it's
enough to compare ``i18nHash`` to decide if component should update.

Using lambda components:

.. code-block:: jsx

   import React from "react"
   import { I18n } from "@lingui/react"

   function LogoutIcon() {
      return (
         <I18n>
            {({ i18n }) => <Icon name="turn-off" ariaLabel={i18n.t`Log out`} />}
         </I18n>
      )
   }

Using components and elements:

.. code-block:: jsx

   import React from "react"
   import { I18n } from "@lingui/react"

   function TranslatedComponent({ i18n }) {
      return <Icon name="turn-off" ariaLabel={i18n.t`Log out`} />
   }

   function RenderingElements() {
      return (
         <I18n>
            <TranslatedComponent />
         </I18n>
      )
   }

   function RenderingComponents() {
      return <I18n>{TranslatedComponent}</I18n>
   }


withI18n
--------

.. js:function:: withI18n(options?)

   :param Object options: Configuration for high-order component
   :param bool update: Subscribe to catalog and activate language updates
   :param bool withHash: Pass unique ``i18nHash`` prop to force underlying PureComponent re-render on catalog and active language update
   :param bool withRef: Returns reference to wrapped instance in ``getWrappedInstance``

:js:func:`withI18n` is a higher-order component which injects ``i18n`` object to
wrapped component. ``i18n`` object is needed when you have to access plain JS
API for translation of JSX props:

.. code-block:: jsx

   import React from "react"
   import { withI18n } from "@lingui/react"

   const LogoutIcon = withI18n()(({ i18n }) => (
     <Icon name="turn-off" ariaLabel={i18n.t`Log out`} />
   ))

.. note:: :js:func:`withI18n` automatically hoists static properties from wrapped component.

i18nMark
--------

.. js:function:: i18nMark(msgId: string)

Mark string as translated text, but don't translate it immediatelly.
This string is extracted to message catalog and can be used in
:component:`Trans`:

.. code-block:: jsx

   const message = i18nMark('Source text')
   const translation = <Trans id={message} />

   // This is the same as:
   // <Trans id="Source text" />

:js:func:`i18nMark` is useful for definition of translations outside
components:

.. code-block:: jsx

   const languages = {
     en: i18nMark('English'),
     fr: i18nMark('French')
   };

   Object.keys(languages).map(language =>
     <Trans key={language} id={languages[language]} />
   );

.. note::

   In development, :js:func:`i18nMark` is an identity function, returning ``msgId``.

   In production, :js:func:`i18nMark` call is replaced with ``msgId`` string.

.. _rendering-translations:

Rendering of translations
=========================

All i18n components render translation as a text without a wrapping tag. This can be
customized in two different ways: globally -- using ``defaultRender`` prop on
:component:`I18nProvider` component; or locally -- using ``render`` prop on i18n
components.

Global Configuration
--------------------

Default rendering component can be set using ``defaultRender`` prop in
:component:`I18nProvider`. The main use case for this is rendering translations
in ``<Text>`` component in React Native.

It's possible to pass in either a string for built-in elements (`span`, `h1`),
React elements or React classes. This prop has the same type as ``render`` prop on
i18n components described below.

Local Configuration
-------------------

============= ==================================== ============================
Prop name     Type                                 Description
============= ==================================== ============================
``className`` string                               Class name to be added to ``<span>`` element
``render``    Element, Component, string, ``null`` Custom wrapper element to render translation
============= ==================================== ============================

``className`` is used only for built-in components (when `render` is string).

When ``render`` is **React.Element** or **string** (built-in tags), it is
cloned with the ``translation`` passed in as its child:

.. code-block:: jsx

   // built-in tags
   <Trans render="h1">Heading</Trans>;
   // renders as <h1>Heading</h1>

   // custom elements
   <Trans render={<Link to="/docs" />}>Link to docs</Trans>;
   // renders as <Link to="/docs">Link to docs</Link>

Using **React.Component** (or stateless component) in ``render`` prop is useful
to get more control over the rendering of translation. Component passed to
``render`` will receive the translation value as a ``translation`` prop:

.. code-block:: jsx

   // custom component
   <Trans render={({ translation }) => <Icon label={translation} />}>
      Sign in
   </Trans>;
   // renders as <Icon label="Sign in" />

``render`` also accepts ``null`` value to render
string without wrapping component. This can be used to override
custom ``defaultRender`` config.

.. code-block:: jsx

   <Trans render={null}>Heading</Trans>;
   // renders as "Heading"

.. _Intl.DateTimeFormat: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/DateTimeFormat
.. _Intl.NumberFormat: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/NumberFormat
