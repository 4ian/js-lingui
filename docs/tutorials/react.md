`lingui-react` provides components which represent ICU MessageFormat in React. It's useful to understand this format, but certainly not required.

:bulb:  All components are exported from the top of `lingui-react` package.

# Install dependencies and configure Babel

1. Install `babel-preset-lingui-react` and `lingui-cli` as development dependencies:

  {% sample lang="js" -%}
  ```bash
  yarn add --dev babel-preset-lingui-react lingui-cli
  # npm install --save-dev babel-preset-lingui-react lingui-cli  
  ```  

2. Install `lingui-react` as runtime dependency:

  {% sample lang="js" -%}
  ```bash
  yarn add --dev lingui-react
  # npm install --save-dev lingui-react
  ```

3. Add `lingui-react` preset to babel config:

  {% sample lang="js" -%}
  ```js
  // .babelrc
  {
    "presets": [
      "env",
      "react",
      "lingui-react"
    ]
  }
  ```

# Setup I18nProvider

Translations are loaded from message catalogs. The topmost 
component which adds messages and language to React component tree is 
[I18nProvider](https://github.com/lingui/js-lingui/wiki/Ref:-React#i18nprovider):

```jsx
import React from 'react'
import { render } from 'react-dom'
import { I18nProvider } from 'lingui-react'

// This is for development only. It contains message parser for
// compiling messages on the fly and language data for all languages
import linguiDev from 'lingui-i18n/dev'

import messages from './locales/en.json'
import App from './App'

render(
    <I18nProvider 
      language="en" 
      messages={{ en: messages }}
      development={linguiDev}>
       <App />
    </I18nProvider>,
    document.getElementById("app")
)
```

:hammer_and_wrench: Under the hood this component adds messages and language to 
context. All `lingui-react` components access messages from this context, so 
it's not necessary to pass them in props manually.

:bulb:  Active language can be managed for example in `redux` store. In that case,
add [I18nProvider](https://github.com/lingui/js-lingui/wiki/Ref:-React#i18nprovider) 
in main component (e.g. `App`) and read `language` from `redux` store. Loading 
of messages dynamically can be achieved in several ways depending on bundler 
(e.g. webpack) and it's not covered in this tutorial. See the
[HowTo: Dynamic loading of languages in Webpack](https://github.com/lingui/js-lingui/wiki/HowTo:-Dynamic-loading-of-languages-with-Webpack)

# Wrap messages in i18n components

The simplest i18n component is [Trans](https://github.com/lingui/js-lingui/wiki/Ref:-React#trans). 
It's used for simple, singular 
translations. It supports both variables and inline components.

Suppose we have a `Header` component, which displays the name of currently
logged in user:

```jsx
import React from 'react'

const Header = ({ username }) => (
  <div>
    <h1>Documentation</h1>
    
    <div className="UserProfile">
        Welcome, <Link to="/profile">{username}</Link>
    </div>
  </div>
)
```

If we want to translate this component, just wrap all messages in
[Trans](https://github.com/lingui/js-lingui/wiki/Ref:-React#trans) component:

```jsx
import React from 'react'
import { Trans } from 'lingui-react'

const Header = ({ username }) => (
  <div>
    <h1><Trans>Documentation</Trans></h1>
    
    <div className="UserProfile">
        <Trans>Welcome, <Link to="/profile">{username}</Link></Trans>
    </div>
  </div>
)
```

This process is completely seamless and doesn't require any other changes
to `Header` component.

**:warning: Only simple variables are supported inside [Trans](https://github.com/lingui/js-lingui/wiki/Ref:-React#trans) component.**
It's not possible to use constants, function calls or object properties.
Any of following examples won't work:

```js
const messages = [
  "Hello",
  "World"
]

<Trans>{messages.length}</Trans>
<Trans>{isEmpty(messages)}</Trans>
<Trans>The answer is {42}</Trans>
```

## :hammer_and_wrench: How the extracted messages are going to look like

Extraction of messages into external message catalogs isn't covered in this
tutorial because it's a responsibility of `lingui-cli` package. However, now it's
probably the best and only time to mention, how extracted messages are going
to look like.

Given the example above, we'll get two messages:

```js
// example of empty message catalog
export default {
  "Documentation": "",
  "Welcome, <0>{username}</0>": ""
}
```

Anything wrapped inside [Trans](https://github.com/lingui/js-lingui/wiki/Ref:-React#trans) component is considered as
a **source message**.

Names of variables are preserved. This is the reason why only simple variables
are allowed - variable names are used as placeholders.

Inline components are replaced with dummy `<0>`, `<1>`, etc. tags. Using 
anonymous tags has several advantages:

- Both built-in (e.g: `span`) and custom (e.g: `Link`) components are supported
- Change of props doesn't require update of translations (e.g. when class name is changed)

# Translations of attributes

Sometimes it's necessary to translate element attributes, like link `title` or
`aria-label`. [Trans](https://github.com/lingui/js-lingui/wiki/Ref:-React#trans) is a React component, but in this case we need
just a text.

It's possible to access low-level i18n API using [WithI18n](https://github.com/lingui/js-lingui/wiki/Ref:-React#withi18n)
high-order component. 
All we need to do is wrap our component and then
we can use `lingui-i18n` API from `i18n` prop:

```jsx
import React from 'react'
import { Trans, WithI18n } from 'lingui-react'

// WithI18n injects `i18n` prop 
const LogoutIcon = WithI18n()(
    ({ i18n }) => <Icon name="turn-off" aria-label={i18n.t`Log out`}/>
)
```

**:warning: Watch the empty braces after `WithI18n`. `WithI18n` expects options
as a first argument and return configured HOC.**

`i18n.t` returns plain text. It's the same API which can be used outside React
context and it's described in `lingui-i18n`

# Plurals, categories, and formats

ICU MessageFormat is very powerful and extensible format. It supports plurals,
categories and date/number formats.

## Plurals

[Plural](https://github.com/lingui/js-lingui/wiki/Ref:-React#plural) component selects the appropriate plural form based on given
`value`. For example in English, there are only two plural forms: singular and
plural. If we want to translate message with plural support, we need to define
both:

```jsx
import React from 'react'
import { Trans } from 'lingui-react'

const Notifications = ({ newMessageCount }) => {
  return (
    <div>
      <Plural 
         value={newMessageCount}
         one="# message"
         other="# messages"
      />
    </div>
  )
}
```

In English, [Plural](https://github.com/lingui/js-lingui/wiki/Ref:-React#plural) will render `1 message`, but `3 messages` depending on
`value`. It means that `one` prop expresses singular while `other` is for plural. 
Plural forms for other than source language are defined in message catalogs.
If we use English as a source language, we'll always use only `one` and `other`
props in [Plural](https://github.com/lingui/js-lingui/wiki/Ref:-React#plural). 

Just to clarify, the message above is extracted as:

`{newMessageCount, plural, one {# message} other {# messages}}`

In Czech, it would be translated as:

`{newMessageCount, plural, one {# zpráva} few {# zprávy} other {# zpráv}}`

This is different compared to `gettext`, which maps singular/plural to
language specific plural forms. In ICU message format, we always have only
one message with all plural forms per language.

:hammer_and_wrench: All possible plural forms are `zero`, `one`, `two`, `few`, 
`many`, `other`.

:bulb: Plural forms for all languages are listed at CLDR page - 
[Language Plural Rules](http://www.unicode.org/cldr/charts/latest/supplemental/language_plural_rules.html).

## Select

Some languages have different forms based on gender or level of politeness.
Such cases are handled using [Select](https://github.com/lingui/js-lingui/wiki/Ref:-React#select) component, which works like
a `switch` statement:

```jsx
import React from 'react'
import { Trans } from 'lingui-react'

const Book = ({ gender }) => {
  return (
    <div>
      <Select 
         value={gender}
         male="His book"
         female="Her book"
         other="Their book"
      />
    </div>
  )
}
```

There's no magic here. [Select](https://github.com/lingui/js-lingui/wiki/Ref:-React#select) uses the form based on `value`.
`other` prop defines default form.

## DateFormat and NumberFormat

Localization isn't just about text, but also about different number and date
formats. [NumberFormat](https://github.com/lingui/js-lingui/wiki/Ref:-React#numberformat) is used for formatting numbers 
(decimals, percents, currencies), while [DateFormat](https://github.com/lingui/js-lingui/wiki/Ref:-React#dateformat) is for
dates:

```jsx
import React from 'react'
import { DateFormat, NumberFormat } from 'lingui-react'

const Invoice = ({ invoiceNumber, datePaid, amount }) => {
  return (
    <div>
      <h1><Trans>Invoice {invoiceNumber}</Trans></h1>

      <Trans>
        Amount <NumberFormat 
          value={amount} 
          format={{ style: 'currency', currency: 'EUR' }} 
        /> paid on <DateFormat value={datePaid} />
      </Trans>
    </div>
  )
}
```

# What's next

Once the app is *internationalized*, it's ready to be *localized*. Localization
process includes extracting messages into message catalog, translating message
catalogs and importing them in the application.

`lingui-cli` provides `extract` command, which creates new or updates existing
message catalogs.

Finally, after the message catalog is translated, it can be loaded back to the
application. [I18nProvider](https://github.com/lingui/js-lingui/wiki/Ref:-React#i18nprovider) expects a dictionary of messages, 
where the key is a message in source language and the value is a message in target language.

Follow up links:

- [Reference: React API](https://github.com/lingui/js-lingui/wiki/Ref:-React)
- [How-to guide: Dynamic loading of languages with Webpack](https://github.com/lingui/js-lingui/wiki/HowTo:-Dynamic-loading-of-languages-with-Webpack) 
