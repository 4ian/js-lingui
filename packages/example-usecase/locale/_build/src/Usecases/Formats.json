{
  "The answer is {answer,number,percent}": {
    "origin": [
      [
        "src/Usecases/Formats.js",
        13
      ]
    ]
  },
  "Today is {now,date}": {
    "origin": [
      [
        "src/Usecases/Formats.js",
        19
      ]
    ]
  }
}