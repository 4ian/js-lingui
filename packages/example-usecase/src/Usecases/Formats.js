import React from 'react'
import { Trans, NumberFormat, DateFormat } from 'lingui-react'

class Formats extends React.Component {
  render () {
    const answer = 0.42
    const now = new Date('4/5/2017')

    return (
      <div>
        <ul>
          <li>
            <Trans className="percent">
              The answer is <NumberFormat value={answer} format="percent" />
            </Trans>
          </li>

          <li>
            <Trans className="date">
              Today is <DateFormat value={now} />
            </Trans>
          </li>
        </ul>
      </div>
    )
  }
}

export default Formats
