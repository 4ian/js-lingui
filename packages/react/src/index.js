// @flow
import { date, number } from "@lingui/core"

import I18nProvider, { I18nCoreConsumer as I18n } from "./I18nProvider"
import Trans from "./Trans"

import createFormat from "./createFormat"
import withI18n from "./withI18n"

const DateFormat = createFormat(date)
const NumberFormat = createFormat(number)

export { withI18n, I18nProvider, I18n, Trans, DateFormat, NumberFormat }
