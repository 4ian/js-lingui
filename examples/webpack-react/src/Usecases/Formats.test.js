// @flow
import * as React from "react"
import { mount } from "enzyme"
import { I18nProvider, setupI18n } from "@lingui/react"

import Formats from "./Formats"

describe("Formats", function() {
  const Component = () => {
    const i18n = setupI18n({ locale: "en" })
    return (
      <I18nProvider i18n={i18n}>
        <Formats />
      </I18nProvider>
    )
  }

  const text = (element, props = {}) => {
    return mount(<Component {...props} />)
      .find(element)
      .first()
      .text()
  }

  it("should render", function() {
    expect(mount(<Component />)).toMatchSnapshot()
  })

  it("should render percent", function() {
    expect(text(".percent")).toEqual("The answer is 42%")
  })

  it("should render date", function() {
    expect(text(".date")).toEqual("Today is 4/5/2017")
  })
})
