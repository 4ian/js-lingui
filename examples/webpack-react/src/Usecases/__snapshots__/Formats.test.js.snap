// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Formats should render 1`] = `
<Component>
  <I18nProvider
    i18n={
      Object {
        "_catalogs": Object {
          "en": Object {
            "localeData": Object {
              "plurals": [Function],
            },
          },
        },
        "_dev": Object {
          "compile": [Function],
          "loadLocaleData": [Function],
        },
        "_didActivate": Array [
          [Function],
        ],
        "_locale": "en",
        "_locales": undefined,
        "_willActivate": Array [],
      }
    }
  >
    <Formats>
      <div>
        <ul>
          <li>
            <Trans$$1
              className="percent"
              id="The answer is {answer,number,percent}"
              values={
                Object {
                  "answer": 0.42,
                }
              }
            >
              The answer is 42%
            </Trans$$1>
          </li>
          <li>
            <Trans$$1
              className="date"
              id="Today is {now,date}"
              values={
                Object {
                  "now": 2017-04-05T11:14:00.000Z,
                }
              }
            >
              Today is 4/5/2017
            </Trans$$1>
          </li>
        </ul>
      </div>
    </Formats>
  </I18nProvider>
</Component>
`;
